internal/integration/missingobjects_test.go

```diff
@@ -0,0 +1,97 @@
+//go:build integration
+
+package integration
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestMissingObjects(t *testing.T) {
+	var info struct {
+		OldOID string `json:"push_from"`
+		NewOID string `json:"push_to"`
+		Ref    string `json:"ref"`
+	}
+	const (
+		remote   = "testdata/missing-objects/remote.git"
+		badPack  = "testdata/missing-objects/bad.pack"
+		infoFile = "testdata/missing-objects/info.json"
+		otherRef = "refs/heads/other"
+	)
+
+	infoJSON, err := os.ReadFile(infoFile)
+	require.NoError(t, err)
+	require.NoError(t, json.Unmarshal(infoJSON, &info))
+
+	origin, err := filepath.Abs(remote)
+	require.NoError(t, err)
+
+	testRepo := t.TempDir()
+	requireRun(t, "git", "init", "--bare", testRepo)
+	requireRun(t, "git", "-C", testRepo, "fetch", origin, info.Ref+":"+info.Ref)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	srp := exec.CommandContext(ctx, "spokes-receive-pack", ".")
+	srp.Dir = testRepo
+	srp.Env = append(os.Environ(),
+		"GIT_SOCKSTAT_VAR_spokes_quarantine=bool:true",
+		"GIT_SOCKSTAT_VAR_quarantine_id=config-test-quarantine-id")
+	srp.Stderr = os.Stderr
+	srpIn, err := srp.StdinPipe()
+	require.NoError(t, err)
+	srpOut, err := srp.StdoutPipe()
+	require.NoError(t, err)
+
+	srpErr := make(chan error)
+	go func() { srpErr <- srp.Run() }()
+
+	bufSRPOut := bufio.NewReader(srpOut)
+
+	refs, _, err := readAdv(bufSRPOut)
+	require.NoError(t, err)
+	assert.Equal(t, refs, map[string]string{
+		info.Ref: info.OldOID,
+	})
+
+	// Try to update the ref that's already there to commit C (but we won't
+	// push its parent and the remote doesn't have the parent either).
+	require.NoError(t, writePktlinef(srpIn,
+		"%s %s %s\x00report-status report-status-v2 side-band-64k object-format=sha1\n",
+		info.OldOID, info.NewOID, info.Ref))
+	// Try to create another ref with a commit that the remote already has.
+	require.NoError(t, writePktlinef(srpIn,
+		"%040d %s %s",
+		0, info.OldOID, otherRef))
+	_, err = srpIn.Write([]byte("0000"))
+	require.NoError(t, err)
+
+	// Send the pack that's missing a commit.
+	pack, err := os.Open("testdata/missing-objects/bad.pack")
+	require.NoError(t, err)
+	if _, err := io.Copy(srpIn, pack); err != nil {
+		t.Logf("error writing pack to spokes-receive-pack input: %v", err)
+	}
+
+	require.NoError(t, srpIn.Close())
+
+	refStatus, unpackRes, _, err := readResult(t, bufSRPOut)
+	require.NoError(t, err)
+	assert.Equal(t, map[string]string{
+		info.Ref: "ng missing necessary objects",
+		otherRef: "ok",
+	}, refStatus)
+	assert.Equal(t, "unpack ok\n", unpackRes)
```

Comment:
Should we accept any ref updates when there are connectivity problems? `git-receive-pack` does, so I think we're maintaining the status quo here. However, accepting any refs here means that the pack file will be copied into the repo, which means we will still get objects with missing ancestors in the repository. The disconnected objects wouldn't be reachable, so they shouldn't cause problems. I did check that `show-index` and `cat-file` don't seem to care about the connectivity problems, so our other checks should work ok.

Reply:
This is why the connectivity check compares against reachable objects (`--not --all --alternate-refs`), not just stopping at objects that are stored in the repository. We are allowed to have objects that reference objects not in the repo, as long as they are not reachable from refs.

Reply:
hmm.. i am wondering if this is an interesting way to hide malicious content without getting it detected by scanning or other tooling (have it mapped in objects that are not reachable from refs).

Reply:
@derrickstolee I'm not so sure that unreachable objects that point at missing objects are safe on our servers. I think that the API allows a reference to be pointed at an arbitrary (including heretofore unreachable) object, and that no transitive connectivity check is done in that situation. I'm pretty sure that the same thing is possible from Ruby.

We strive not to create objects on our servers that are not closed under reachability, for example by checking in libgit2 that a newly-created object only points at existing objects. Before we did this reliably, it led to occasional corruption.

Reply:
@mhagger I agree that we should avoid unreachable objects that point to missing objects. I just know that this is a possible/expected data shape on our servers (and we need to keep that in mind when doing changes in `github/git`).

@AshwinMohan86 The way to fix this is to scan all objects in the pack, not necessarily the ref tips that are being updated. But I thought the point of this work was to mimic the behavior of `git-receive-pack` bug-for-bug and not to introduce a change at this point.